pricegraph/src/graph/bellman_ford.rs

```diff
@@ -0,0 +1,149 @@
+///
+/// The orginal source can be found here:
+/// https://docs.rs/petgraph/0.5.0/src/petgraph/algo/mod.rs.html#745-792
+pub fn search<G>(g: G, source: G::NodeId) -> Result<Paths<G>, NegativeCycle<G>>
+where
+    G: NodeCount + IntoNodeIdentifiers + IntoEdges + NodeIndexable,
+    G::EdgeWeight: FloatMeasure,
+{
+    let mut predecessor = vec![None; g.node_bound()];
+    let mut distance = vec![<_>::infinite(); g.node_bound()];
+
+    let ix = |i| g.to_index(i);
+
+    distance[ix(source)] = <_>::zero();
+    // scan up to |V| - 1 times.
+    for _ in 1..g.node_count() {
+        let mut did_update = false;
+        for i in g.node_identifiers() {
+            for edge in g.edges(i) {
+                let i = edge.source();
+                let j = edge.target();
+                let w = *edge.weight();
+                if distance[ix(i)] + w < distance[ix(j)] {
+                    distance[ix(j)] = distance[ix(i)] + w;
+                    predecessor[ix(j)] = Some(i);
+                    did_update = true;
+                }
+            }
+        }
+        if !did_update {
+            break;
+        }
+    }
+
+    // check for negative weight cycle
+    for i in g.node_identifiers() {
+        for edge in g.edges(i) {
+            let j = edge.target();
+            let w = *edge.weight();
+            if distance[ix(i)] + w < distance[ix(j)] {
+                // NOTE: The following 4 lines are the only modifications made
+                // to the original algorithm and where originally:
+                // ```diff
+                // -//println!("neg cycle, detected from {} to {}, weight={}", i, j, w);
+                // -return Err(NegativeCycle(()));
+                // ```
+                predecessor[ix(j)] = Some(i);
+                let cycle = find_cycle::<G>(g, j, &predecessor)
+                    .expect("negative cycle not found after being detected");
+                return Err(NegativeCycle(cycle));
+            }
+        }
+    }
+
+    Ok((distance, predecessor))
+}
+
+/// Finds a cycle and returns a vector representing a path along the cycle,
+/// ending that is the predecessor of the starting node.
+///
+/// Returns `None` if no such cycle can be found.
+fn find_cycle<G>(
+    g: G,
+    start: G::NodeId,
+    predecessor: &[Option<G::NodeId>],
+) -> Option<Vec<G::NodeId>>
+where
+    G: NodeCount + NodeIndexable,
+{
+    let ix = |i| g.to_index(i);
+
+    // NOTE: First find a node that is actually on the cycle, this is done
+    // because a negative cycle can be detected on any node connected to the
+    // cycle and not just nodes on the cycle itself.
+    let mut visited = vec![false; g.node_bound()];
+    let mut current = start;
+    visited[ix(current)] = true;
+    loop {
+        let next = predecessor[ix(current)]?;
+        if visited[ix(next)] {
+            break;
+        }
+        visited[ix(next)] = true;
+        current = next;
```

Comment:
nit: I think only one of current/next should be needed as in the loop below.

Reply:
Nice find.

Reply:
:eagle: :eye: